homeassistant/components/roborock/roborock_storage.py

```diff
@@ -0,0 +1,152 @@
+"""Roborock storage."""
+
+import dataclasses
+import logging
+import os
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.storage import Store
+from homeassistant.util import dt as dt_util
+
+from .const import DOMAIN
+from .models import RoborockMapInfo
+
+STORAGE_VERSION = 1
+_LOGGER = logging.getLogger(__name__)
+MAP_PATH = f"{DOMAIN}/maps"
+MAP_UPDATE_FREQUENCY = 3600  # Only save the map once every hour.
+
+
+@dataclasses.dataclass
+class RoborockMapEntry:
+    """Describe a map entry stored on disk."""
+
+    map_name: str
+    time: float
+
+
+class RoborockStorage:
+    """Store Roborock data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry_id: str,
+    ) -> None:
+        """Initialize RoborockStorage."""
+        self._hass = hass
+        self._data: dict[str, RoborockMapEntry] = {}
+        self._entry_id = entry_id
+        self._store: Store = Store(hass, STORAGE_VERSION, f"{DOMAIN}.{self._entry_id}")
+        self._map_info: dict[str, dict[int, RoborockMapInfo]] = {}
+
+    def _should_update(self, map_entry: RoborockMapEntry | None) -> bool:
+        return (
+            map_entry is None
+            or dt_util.utcnow().timestamp() - map_entry.time > MAP_UPDATE_FREQUENCY
+        )
+
+    def _get_map_filename(self, map_name: str) -> str:
+        return self._hass.config.path(f"{MAP_PATH}/{self._entry_id}/{map_name}")
+
+    def exec_load_maps(self, map_names: list[str]) -> list[bytes | None]:
+        """Load map content. Should be called in executor thread."""
```

Comment:
i think it would be safer to separate the map loading disk i/o from updating the python object state (`_data`). That is, it seems like the save and load can happen from different executor threads.

Perhaps provide async interfaces and let those update the object state and defer to the executor threads for I/O only. 

When writing, i think its fine to only update `_data` after the map write succeeds rather than writing and handling a rollback case on error.

Reply:
> When writing, i think its fine to only update _data after the map write succeeds rather than writing and handling a rollback case on error.

Seems to be what i do - so I must have changed this in the past - resolving this conversation.

Reply:
What about the rest of the comment? I don't see that this was fixed, so I can't tell if you disagree with these proposed updates or if it was not clear what I was referring to about the safety issues here.  Here is a spelled out example:

Two different entities could have active requests to `async_image` in parallel. Each one reloads map data, calls `async_save_map` in parallel, then both run a partial update of `async_save_map` which both run a partial update in `async_save_maps` which can run two different executors both updating `_data`.

Stepping back: The general theme of my comments is about defining more clear boundaries between the classes so that they are safer and easier to reason about. Partial update support is very complicated to implement correctly, which is why I am suggesting separating the I/O from the data structure management.